ncm-metaconfig/src/main/perl/metaconfig.pm

```diff
@@ -223,16 +223,61 @@ sub process_actions
     }
 }
 
+# Run $service shell command of $type (ie a string) (if defined).
+#   $msg is a reporting prefix
+#   When $input is not undef, pass it on stdin
+# Return 1 on success, undef otherwise.
+sub run_shell_command
+{
+    my ($self, $commands, $type, $input) = @_;
+
+    my $command = $commands->{$type};
+    if ($command) {
+        $self->debug(1, "Going to run $type command '$command'");
+
+        my ($err, $out);
+        my %opts = (
+            shell => 1,
```

Comment:
@gombasg we can also do that. at some point it might make more sense to store this command/action registry somewhere central; and have more direct access to the commands from NCM::NCD or CAF::Proces.
i can start with putting where you suggest it, and see how useful it is.

(and obviously, commands need to be registered with `final` keyword.)

Reply:
I'd still vote for shell=>0. Looking at the docs, this should work just fine if we want the input to be strings?

```
my $cmd = CAF::Process->new(["ls -lh"], log => $self);
```

One of the reasons this structure is nice is that I can also nullify or replace any command I don't like with a central dict update:

```
'metaconfig/commands/not-for-me' = '/bin/true';
```

without having to loop through all `services` to find the right one.

I could potentially also do that conditionally using Aquilon's build status to e.g. avoid running commands that I know won't work during anaconda and but allowing them to run fine once we've left buildstatus=build.

(yeah, that's a bit ugly but potentially less so than the `cmd||/bin/true` hacks we have at the moment)

Reply:
Just spotted this use of shell commands in our code base:
```
/usr/sbin/alternatives --set mta /usr/sbin/sendmail.sendmail; systemctl restart sendmail.service
```

But then with this code, I think we could still do that? The alternatives command would be run by the changed or post action, and then the restart would happen via the standard daemons attribute.

Reply:
the daemons are only triggered by a change, you can also run the alternatives in the pre section.
when is it triggered now? only on change?

Reply:
Yep, on change only via file copy.

Reply:
in that case, yes this would work (but only because you are lucky enough that last command is a daemon action).

Reply:
Right, but my point is that a lot of where we currently have shell "tricks" could be re-thought to be cleaner. But if we give people a $shell=>1 crutch they'll just port directly across from filecopy without trying to find a better way. 

What we found with file copy is that once sites are using shell=1, there's now way to back out of it without risk of breakage. So it's much better to start with shell=0 and then only turn it on if we find a proper use case that can't be solved any other way.